drakrun/drakrun/storage.py

```diff
@@ -0,0 +1,261 @@
+                    os.path.join(self.zfs_tank_name, f"vm-{vm_id}"),
+                ],
+                check=True,
+            )
+
+            for _ in range(120):
+                if not os.path.exists(vm_zvol):
+                    time.sleep(0.1)
+                else:
+                    break
+            else:
+                raise RuntimeError(f"Failed to see {vm_zvol} created after executing "
+                                   "zfs clone command.")
+
+            subprocess.run(["zfs", "snapshot", vm_snap], check=True)
+
+        subprocess.run(["zfs", "rollback", vm_snap], check=True)
+
+    @contextlib.contextmanager
+    def vm0_root_as_block(self) -> Generator[str, None, None]:
+        # workaround for not being able to mount a snapshot
+        base_snap = shlex.quote(os.path.join(self.zfs_tank_name, "vm-0@booted"))
+        tmp_snap = shlex.quote(os.path.join(self.zfs_tank_name, "tmp"))
+        try:
+            subprocess.check_output(f"zfs clone {base_snap} {tmp_snap}", shell=True)
+        except subprocess.CalledProcessError:
+            raise RuntimeError("Failed to clone temporary zfs snapshot")
+
+        volume_path = os.path.join("/", "dev", "zvol", self.zfs_tank_name, "tmp-part2")
+        # Wait for 60s for the volume to appear in /dev/zvol/...
+        for _ in range(60):
+            if os.path.exists(volume_path):
+                break
+            time.sleep(1.0)
+        else:
+            raise RuntimeError(f"ZFS volume not available at {volume_path}")
+
+        yield volume_path
+
+        subprocess.check_output(f"zfs destroy {tmp_snap}", shell=True)
+
+
+class Qcow2StorageBackend(StorageBackendBase):
+    """ Implements storage backend based on QEMU QCOW2 image format """
+
+    def __init__(self, install_info: InstallInfo):
+        super().__init__(install_info)
+        self.check_tools()
+
+    @staticmethod
+    def check_tools():
+        """ Verify existence of qemu-img """
+        try:
+            subprocess.check_output("qemu-img --version", shell=True)
+        except subprocess.CalledProcessError:
+            raise RuntimeError("Failed to determine qemu-img version. "
+                               "Make sure you have qemu-utils installed.")
+
+    def initialize_vm0_volume(self, disk_size: str):
+        try:
+            subprocess.check_output(
+                " ".join(
+                    [
+                        "qemu-img",
+                        "create",
+                        "-f",
+                        "qcow2",
+                        os.path.join(LIB_DIR, "volumes", "vm-0.img"),
+                        shlex.quote(disk_size),
+                    ]
+                ),
+                shell=True,
+            )
+        except subprocess.CalledProcessError:
+            raise RuntimeError("Failed to create a new volume using qemu-img.")
+
+    def snapshot_vm0_volume(self):
+        # We'll be using vm-0.img as backing storage
```

Comment:
I'd add a log here (might come in handy in the future while debugging).

Reply:
+1 to that, however how do we do it cleanly with karton? AFAIR `import logging` won't work. Or will it?

Reply:
I think you can do `logging.getLogger("karton")` (or w/e kartonik you want).

Reply:
OK, leaving that for now. Added a TODO.